db/virtual_table: Streaming tables for large data + describe_ring example table

jul-stas opened the pull request:
This is the 2nd PR in series with the goal to finish the hackathon project authored by @tgrabiec, @kostja, @amnonh and @mmatczuk (improved virtual tables + function call syntax in CQL). This one introduces a new implementation of the virtual tables, the streaming tables, which are suitable for large amounts of data.

This PR was created by @jul-stas and @StarostaGit

tgrabiec commented:
Looks like my comments on the previous PR were not addressed: https://github.com/scylladb/scylla/pull/8464#pullrequestreview-691740349

jul-stas commented:
> Looks like my comments on the previous PR were not addressed: [#8464 (review)](https://github.com/scylladb/scylla/pull/8464#pullrequestreview-691740349)

Aboslutely, it's just a draft!

jul-stas commented:
@tgrabiec I addressed the comments you mentioned, but this one only partially: https://github.com/scylladb/scylla/pull/8464#discussion_r657937937
That's because some code paths in tests call `ms.make_reader()` with the default `fwd_mr = mutation_reader::forwarding::yes`.

jul-stas commented:
@tgrabiec ping

tgrabiec commented on test/lib/mutation_source_test.cc:
```diff
@@ -1636,25 +1636,28 @@ void test_reader_conversions(populate_fn_ex populate) {
     test_slicing_with_overlapping_range_tombstones(populate);
+
+    if (with_partition_range_forwarding) {
+        test_range_tombstones_v2(populate);
```
The queue reader will already throw bad_function_call.

jul-stas replied:
Done. 2 tests really relied on forwarding, so I left the flag `with_partition_range_forwarding`.

jul-stas commented on flat_mutation_reader.cc:
```diff
@@ -971,6 +971,96 @@ make_flat_mutation_reader_from_fragments(schema_ptr schema, reader_permit permit
+
+                        case mutation_fragment::kind::range_tombstone:
+                            auto&& rt = mf.as_range_tombstone();
+                            if (_ranges_walker->advance_to(rt.position(), rt.end_position())) {
```
Done. Although I used `range_tombstone_splitter`

jul-stas commented:
@tgrabiec I addressed your last comments, would you mind reiterating?

jul-stas commented:
@avikivity Tomek is, apparently, on vacation and this PR is hanging for quite some time.
Would you mind having a look?

avikivity commented on mutation_reader.hh:
```diff
@@ -136,6 +136,8 @@ flat_mutation_reader make_filtering_reader(flat_mutation_reader rd, MutationFilt
     return make_flat_mutation_reader<filtering_reader<MutationFilter>>(std::move(rd), std::forward<MutationFilter>(filter));
 }
 
+flat_mutation_reader make_slicing_filtering_reader(flat_mutation_reader, const dht::partition_range&, const query::partition_slice&);
```
Please add a comment explaining what a slicing filtering reader is.

avikivity commented on flat_mutation_reader.cc:
```diff
@@ -971,6 +971,105 @@ make_flat_mutation_reader_from_fragments(schema_ptr schema, reader_permit permit
+                return _rd.fill_buffer(timeout).then([this, consume_fn] {
+                    while (!_rd.is_buffer_empty()) {
+                        auto mf = _rd.pop_mutation_fragment();
+                        switch (mf.mutation_fragment_kind()) {
```
How about writing this as a coroutine? do_until becomes `while`, and .then becomes `co_await`. It's not strictly required but it's easy and improves readability.

jul-stas replied:
Done

avikivity commented on flat_mutation_reader.cc:
```diff
@@ -971,6 +971,105 @@ make_flat_mutation_reader_from_fragments(schema_ptr schema, reader_permit permit
+        virtual future<> fill_buffer(db::timeout_clock::time_point timeout) override {
+            const auto consume_fn = [this] (mutation_fragment mf) {
+                push_mutation_fragment(std::move(mf));
+            };
```
This wrapper adds nothing, AFAICT.

jul-stas replied:
I wrote the wrapper because this appears below and looks awkward:
```
_splitter->flush(mf.position(), std::bind(&reader::push_mutation_fragment, this)...
```

avikivity commented on mutation_reader.cc:
```diff
@@ -2052,7 +2052,13 @@ class queue_reader final : public flat_mutation_reader::impl {
+                return next_partition();
+            });
+        }
+        return make_ready_future<>();
```
@denesb please review this

the db::no_timeout seems wrong, but it's a problem in the API itself. Will go away when we remove timeout from fill_buffer.

Just reading stuff from queue_reader until we exhaust the partition is wasteful, but that's a queue_reader limitation, there is limited interaction with the producer. Do we really need queue_reader? I think the virtual reader could be a from-scratch reader and then communicate with execute() to ask it to skip. Anyway this is okay for the first iteration.

avikivity commented on db/system_keyspace.cc:
```diff
@@ -1844,6 +1845,95 @@ class nodetool_status_table : public memtable_filling_virtual_table {
+            }
+
+            co_await result.emit_partition_end();
+        }
```
Nice and simple!

avikivity commented:
Looks generally okay.

denesb commented:
You basically re-invented the generator, except in a much more convoluted way. I think it'd be worth investing into a generating reader, which takes a coroutine which can generate fragments with `co_yield`, kinda like the new sstable reader does (https://github.com/scylladb/scylla/pull/9002).
We could solve the problem with `next_partition()` and `fast_forward_to()` by returning something from `co_yield(mutation_fragment{})` which tells us to skip over a partition or a range.

denesb commented:
That said I think it is fine to merge as-is, but this is more complicated than it needs to be, but that can be solved in a follow-up.

jul-stas commented:
@avikivity @denesb Reiterated and rebased to master, so ping

haaawk commented:
@avikivity ping
